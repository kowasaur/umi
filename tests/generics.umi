T identity[T](T thing) {
    T temp = thing
    temp
}

class Parent {
    Parent() {}
    Void talk() { print("Hello") }
}

class Child : Parent {
    Child() {}
    Void talk() { print("bazinga!") }
}

# I recognise that this example is stupid because the generic is completely unnecessary
Void saySomething[T : Parent](T person) {
    person.talk()
}


Void main() {
    Int i = identity[Int](1)
    Bool b = identity[Bool](true)
    print(i)
    print(b)

    saySomething[Parent](Parent())
    saySomething[Child](Child())
}
