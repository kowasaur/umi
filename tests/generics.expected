1
True
Hello
bazinga!
